src/JBSwapTerminal.sol

```diff
@@ -165,75 +139,83 @@ contract JBSwapTerminal is JBPermissioned, Ownable, IJBTerminal, IJBPermitTermin
 
     /// @notice Return all the accounting contexts for a specified project ID.
     /// @dev    This includes both project-specific and generic accounting contexts, with the project-specific contexts
     ///         taking precedence.
     /// @param projectId The ID of the project to get the accounting contexts for.
-    /// @return An array of `JBAccountingContext` containing the accounting contexts for the project ID.
-    function accountingContextsOf(uint256 projectId) external view override returns (JBAccountingContext[] memory) {
+    /// @return contexts An array of `JBAccountingContext` containing the accounting contexts for the project ID.
+    function accountingContextsOf(uint256 projectId) external view override returns (JBAccountingContext[] memory contexts) {
+        // Keep a reference to the tokens that have a known context for the project.
         address[] memory projectTokenContexts = _tokensWithAContext[projectId];
+
+        // Keep a reference to the default tokens that have a known context.
         address[] memory genericTokenContexts = _tokensWithAContext[DEFAULT_PROJECT_ID];
 
-        JBAccountingContext[] memory contexts =
+        // Combine the two.
+        contexts =
             new JBAccountingContext[](projectTokenContexts.length + genericTokenContexts.length);
-        uint256 actualLength = projectTokenContexts.length;
+
+        // Keep a reference to the number of project-specific contexts.
+        uint256 numberOfProjectTokenContexts = projectTokenContexts.length;
 
         // include all the project specific contexts
-        for (uint256 i = 0; i < projectTokenContexts.length; i++) {
+        for (uint256 i; i < numberOfProjectTokenContexts; i++) {
             contexts[i] = _accountingContextFor[projectId][projectTokenContexts[i]];
         }
 
+        // Keep a reference to the number of generic contexts.
+        uint256 numberOfGenericTokenContexts = genericTokenContexts.length;
+
         // add the generic contexts, iff they are not defined for the project (ie do not include duplicates)
-        for (uint256 i = 0; i < genericTokenContexts.length; i++) {
+        for (uint256 i; i < numberOfGenericTokenContexts; i++) {
             bool skip;
 
-            for (uint256 j = 0; j < projectTokenContexts.length; j++) {
+            for (uint256 j; j < numberOfProjectTokenContexts; j++) {
                 if (projectTokenContexts[j] == genericTokenContexts[i]) {
                     skip = true;
                     break;
                 }
             }
 
             if (!skip) {
-                contexts[actualLength] = _accountingContextFor[DEFAULT_PROJECT_ID][genericTokenContexts[i]];
-                actualLength++;
+                contexts[numberOfProjectTokenContexts + i] = _accountingContextFor[DEFAULT_PROJECT_ID][genericTokenContexts[i]];
             }
         }
-
-        // Downsize the array to the actual length, if needed
-        if (actualLength < contexts.length) {
```

Comment:
@drgorillamd i didnt see a need for this. am i missing something?

Reply:
Hmm I think this is needed.

In the case that the default context has a context for `TokenA` but the project also has a context for `TokenA` the size of the `contexts` array would be `2`, but since they are the same token only one of the contexts is included. Which leaves the last index in the array uninitialized. This mstore changes the length of the array to be `1`  which causes the last (uninitialized) value to no longer be part of the array.

Reply:
oh yes true. reverting